src/lib.rs

```diff
@@ -23,42 +25,36 @@ use std::io::Cursor;
 ///     trialValue, = unpack('>Q',hashlib.sha512(hashlib.sha512(pack('>Q',nonce) + initialHash).digest()).digest()[0:8])
 ///
 /// This value should be less than or equal to the target
-fn bmpow(target: u64, hash: [u8; 64], starter: u64, stepsize: u64, chan_out: Sender<u64>) {
-    let mut nonce: u64 = starter;
-    let mut algoresult;
-
-    loop {
-        let mut wtr = vec![];
-        let mut result: [u8; 64] = [0; 64];
-        let mut hasher_inner = Sha512::new();
-        let mut hasher_outer = Sha512::new();
-
-        nonce += stepsize;
-        match wtr.write_u64::<BigEndian>(nonce) {
-            Ok(_) => {},
-            Err(e) => { println!("error writing endian: {}", e) },
+fn bmpow(target: u64, hash: &[u8; 64], starter: u64, stepsize: u64, cancelled: &AtomicBool) -> u64 {
+    for i in 0.. {
+        if cancelled.load(Relaxed) {
+            return 0;
         }
-        hasher_inner.input(&wtr);
-        hasher_inner.input(&hash);
-
-        hasher_inner.result(&mut result);
-        hasher_outer.input(&result);
-
-        let mut result_outer: [u8; 64] = [0; 64];
-        hasher_outer.result(&mut result_outer);
-
-        let mut r2 = vec![0; 64];
-        hasher_outer.result(&mut r2);
-        let mut rdr = Cursor::new(r2);
-        // Converting from BigEndian to the endinannes of the system
-        algoresult = rdr.read_u64::<BigEndian>().unwrap();
-        if algoresult < target {
-            chan_out.send(nonce).unwrap();
-            return;
+        let nonce = starter + i * stepsize;
+        if hash_nonce(nonce, &hash) < target {
+            return nonce;
         }
     }
+    unreachable!();
+}
+
+fn hash_nonce(nonce: u64, hash: &[u8; 64]) -> u64 {
+    let nonce_bytes: [u8; 8] = unsafe {mem::transmute(nonce.to_be())};
```

Comment:
If you want to avoid unsafe, you could also do the translation manually, something like:
```rust
let nonce_bytes = [(nonce >> 56) as u8, ((nonce >> 48) & 0xff) as u8, ((nonce >> 40) & 0xff) as u8, ...];
```